### PR TITLE
Force calibration: Drag correction

### PR DESCRIPTION
**Pull request by JoepVanlier**
**Why this PR?**
Both for distances very close to the surface as well as axial force, the hydrodynamic model is less than suitable.
In these cases it is recommended to just use a simple Lorentzian, but take the distance to the nearest surface into account through the drag coefficient.

The effective drag coefficient at a particular distance from the surface is given by:
`gamma(l) = gamma_bulk * height_factor(l)`

This PR adds this option.

**More info**
When the hydrodynamic flag is set to off, but a height is specified, it will apply a correct correction on the drag force.
Note that the correction is different for axial and lateral force! For this an extra flag was added to the calibration models named `axial`.

Note that in the future, we will rely on carrying over drag coefficients from lateral to axial calibrations. The way I set up the models in this PR, this is all handled by the calibration models themselves. When doing any sort of calibration, the model will always return "bulk" values. Then the models only responsibility is to forward correct (and they don't need to know where the bulk value came from). This decouples the from and target model.

Implicitly, it would amount to the same as calculating a local drag coefficient for lateral, uncorrecting it with Faxen's law and then correcting the result with Brenner's law. While we explicitly test the lateral -> axial behaviour in a test, we do not document this yet, since a good API for this procedure is still t.b.d.

I have also added a test to specifically test the transfer of drag coefficients from lateral to axial calibration.

**Recommended reading (if you are interested)**: 
Schäffer, E., Nørrelykke, S. F., & Howard, J. (2007). Surface forces and drag coefficients of microspheres near a plane surface measured with optical tweezers. Langmuir, 23(7), 3654-3665.
https://publications.mpi-cbg.de/Sch%C3%A4ffer_2007_894.pdf

### CODE REVIEW

File: lumicks/pylake/force_calibration/tests/test_active_calibration.py [hunk 1]
```diff
@@ -139,3 +139,63 @@ def test_bias_correction():
 
     assert fit_biased.params["Bias correction"].value is False
     assert fit_debiased.params["Bias correction"].value is True
+
+
+def test_faxen_correction_active():
+    """Active calibration should barely be affected by surface corrections for the drag coefficient.
+    However, the interpretation of gamma_ex, which may be carried over to the other calibration
+    *is* important, so this should be covered by a specific test."""
```
Comment (dean0x7d) [inline lumicks/pylake/force_calibration/tests/test_active_calibration.py:147]: The test docstrings are excellent! Great to have the purpose of the test explained!

File: lumicks/pylake/force_calibration/calibration_models.py [hunk 2]
```diff
@@ -158,13 +159,15 @@ class PassiveCalibrationModel:
         Enable hydrodynamic correction.
     distance_to_surface : float, optional
         Distance from bead center to the surface [um].
-        Currently not supported for models without hydrodynamics.
+        None uses the approximation valid for deep in bulk.
```
Comment (dean0x7d) [inline lumicks/pylake/force_calibration/calibration_models.py:162]: It took me a while to understand this sentence. Not sure I have a good suggestion to rephrase it, but maybe you do. (It's also good to use backticks to distinguish `None` as a keyword.)

File: lumicks/pylake/force_calibration/calibration_models.py [hunk 3]
```diff
@@ -228,18 +241,28 @@ def __init__(
                 else self.distance_to_surface * 1e-6,  # um => m
             )
         else:
-            if self.distance_to_surface:
-                raise NotImplementedError(
-                    "Using a distance to the surface is currently not supported for models without "
-                    "hydrodynamic effects."
+            if distance_to_surface:
+                args = [distance_to_surface * 1e-6, bead_diameter * 1e-6 / 2.0]
```
Comment (dean0x7d) [inline lumicks/pylake/force_calibration/calibration_models.py:245]: Tuples are a better default for immutable data. They better convey the intent and are also faster (not that performance matters here, but it's a better default).
```suggestion
                args = (distance_to_surface * 1e-6, bead_diameter * 1e-6 / 2.0)
```

**Review by dean0x7d — APPROVED** [review]
This looks pretty good to me! I only have a couple of minor comments below. And I assume that you'd want to add something to the changelog.

File: lumicks/pylake/force_calibration/tests/test_hydro.py [hunk 4]
```diff
@@ -434,7 +435,33 @@ def test_distance_to_surface_input(integration_test_parameters):
...
     ActiveCalibrationModel(distance_to_surface=None, hydrodynamically_correct=True, **pars)
 
+
+def test_invalid_model():
     with pytest.raises(NotImplementedError):
-        ActiveCalibrationModel(
-            distance_to_surface=0.5 * 1.51, hydrodynamically_correct=False, **pars
-        )
+        PassiveCalibrationModel(1.0, hydrodynamically_correct=True, axial=True)
+
```
[the remark this answers is not in the document]
Comment (JoepVanlier) [reply lumicks/pylake/force_calibration/tests/test_hydro.py:442]: Yeah, good point. I tried to make a habit out of doing this. Not sure why I overlooked it this time. Thanks for pointing it out.

File: lumicks/pylake/force_calibration/calibration_models.py (hunk 2 above)
Comment (JoepVanlier) [reply lumicks/pylake/force_calibration/calibration_models.py:162]: Yeah, fair point. I'll rephrase it to:
```python

File: lumicks/pylake/force_calibration/calibration_models.py (hunk 3 above)
Comment (JoepVanlier) [reply lumicks/pylake/force_calibration/calibration_models.py:245]: Good point!

File: lumicks/pylake/force_calibration/calibration_models.py [hunk 5]
```diff
@@ -158,13 +159,16 @@ class PassiveCalibrationModel:
         Enable hydrodynamic correction.
     distance_to_surface : float, optional
         Distance from bead center to the surface [um].
-        Currently not supported for models without hydrodynamics.
+        When specifying `None`, the model will use an approximation which is only suitable for
+        measurements performed deep in bulk.
```
Comment (lpingas) [inline lumicks/pylake/force_calibration/calibration_models.py:163]: Now that this is supported without hydro, the check for `hydrodynamically_correct` on line 192 has to be removed.

File: lumicks/pylake/force_calibration/calibration_models.py [hunk 6]
```diff
@@ -209,6 +215,14 @@ def __init__(
...
+                raise NotImplementedError(
+                    "No hydrodynamically correct axial force model is currently available."
+                )
+
+            # The hydrodynamically correct model already accounts for the effect of a nearby wall.
+            # Here the drag coefficient in the model represents the bulk drag coefficient.
+            self._drag_correction_factor = 1
             # This model is only valid up to l/R < 1.5 [6] so throw in case that is violated.
             if distance_to_surface and distance_to_surface / (self.bead_diameter / 2) < 1.5:
                 raise ValueError(
```
Comment (lpingas) [inline lumicks/pylake/force_calibration/calibration_models.py:228]: Not really related to this PR, but I noticed during testing that this message could be a little bit clearer:
It raises
``` 
"This model is only valid for distances to the surface ..."
```
but it could be something like
```
"A hydrodynamically correct model is only valid for distances to the surface ..."
```

In this way, the user can figure out that hydro can be disabled if they are working so close to the surface.

Comment (dean0x7d) [reply lumicks/pylake/force_calibration/calibration_models.py:228]: The error message could also include explicit troubleshooting advice: "If you are working very close to the surface, turn off the hydrodynamic correction."

File: lumicks/pylake/force_calibration/calibration_models.py (hunk 5 above)
Comment (JoepVanlier) [reply lumicks/pylake/force_calibration/calibration_models.py:163]: Damn! Good catch!

File: lumicks/pylake/force_calibration/calibration_models.py (hunk 6 above)
Comment (JoepVanlier) [reply lumicks/pylake/force_calibration/calibration_models.py:228]: Good ideas. I've combined the two suggestions into:
```python
                    "The hydrodynamically correct model is only valid for distances to the surface "
                    "larger than 1.5 times the bead radius. For distances closer to the surface, "
                    "turn off the hydrodynamically correct model."
```

**Review by rpauszek — APPROVED** [review]
Cool, looks great. just 2 very minor comments about documentation

File: docs/tutorial/force_calibration.rst [hunk 7]
```diff
@@ -186,6 +186,49 @@ Note that when `rho_sample` and `rho_bead` are omitted, values for water and pol
...
+        + \frac{1R^5}{5l^5}
+        + \frac{7R^{11}}{200l^{11}}
+        - \frac{1R^{12}}{25l^{12}}
+    }
+
+This model deviates less than 0.1% from Brenner's exact formula for :math:`l/R >= 1.1` and less than 0.3% over the entire range of :math:`l` :cite:`schaffer2007surface`:.
+
+This model can be used in Pylake by specifying `axial=True`::
+
+    force_model = lk.PassiveCalibrationModel(bead_diameter, hydrodynamically_correct=False, distance_to_surface=1e-6, axial=True)
```
[the remark this answers is not in the document]
Comment (JoepVanlier) [reply docs/tutorial/force_calibration.rst:230]: Right. Good point. I'll leave `hydrodynamically_correct=False` out entirely. It's not relevant here.

File: lumicks/pylake/force_calibration/calibration_models.py [hunk 8]
```diff
@@ -228,18 +238,28 @@ def __init__(
...
         physical_spectrum = self._passive_power_spectrum_model(f, fc, diffusion_constant)
         return physical_spectrum * self._filter(f, *filter_params)
 
+    @property
+    def _drag(self):
+        """Returns the corrected drag coefficient
+
+        Note that when using the hydrodynamic model, no correction is necessary.
+        """
+        return self.drag_coeff * self._drag_correction_factor
```
[the remark this answers is not in the document]
Comment (JoepVanlier) [reply lumicks/pylake/force_calibration/calibration_models.py:261]: Yeah, it's a good point. In ten years we'll be thankful for that comment. I added the following:
```
Note that the hydrodynamic model already has a surface dependent drag coefficient baked in and therefore no correction is necessary (in this case self._drag_correction_factor is set to 1).
```
